docs/hugo/content/design/ADR-2024-10-OneOf-Resources/_index.md

```diff
@@ -0,0 +1,253 @@
+}
+
+class StudentProperties {
+    Class string
+    Grade string
+    StartDate date
+    FinishDate optional<date>
+}
+
+
+class TutorProperties {
+    Class string
+    Timeslot string
+    StartDate date
+    FinishDate optional<date>
+}
+
+class TeacherProperties {
+    Subject string
+    Tenure bool
+    StartDate date
+    FinishDate optional<date>
+}
+
+class MarkerProperties {
+    Exam string
+    Strictness string
+    StartDate date
+    FinishDate optional<date>
+}
+
+Role *-- StudentProperties : Student
+Role *-- TutorProperties : Tutor
+Role *-- TeacherProperties : Teacher
+Role *-- MarkerProperties : Marker
+```
+
+### Background: Resource Structure
+
+When we generate the ARM types for submission to Azure Resource Manager, we require each resource to have a `Name` property. 
+
+If that name is missing, we get the following error:
+
+```
+ error generating code: 
+ failed to execute stage 38: 
+ Create types for interaction with ARM: 
+ unable to create arm resource spec definition for resource <id>: 
+ resource spec doesn't have "Name" property"
+```
+
+### The Problem
+
+When importing resources from `Microsoft.Kusto` we have, for the first time, a resource using a _One-Of_ as a part of it's definition right at the root. There are two flavours of the `Database` resource, one for read/write, and another for following.
+
+``` mermaid
+classDiagram
+
+class ClustersDatabase_Spec {
+    <<oneof>>
+    Name string
+}
+
+class ReadWriteDatabase
+class ReadOnlyFollowingDatabase
+
+ClustersDatabase_Spec --> ReadWriteDatabase : ReadWrite
+ClustersDatabase_Spec --> ReadOnlyFollowingDatabase : ReadOnlyFollowing
+```
+
+When the OneOf is rendered according to our current rules, we end up with this structure:
+
+``` mermaid
+classDiagram
+
+class ClustersDatabase_Spec {
+    <<oneof>>
+    ReadWrite ReadWriteDatabase
+    ReadOnlyFollowing ReadOnlyFollowingDatabase
+}
+
+class ReadWriteDatabase {
+    Name string
+}
+
+class ReadOnlyFollowingDatabase {
+    Name string
+}
+
+ClustersDatabase_Spec *-- ReadWriteDatabase
+ClustersDatabase_Spec *-- ReadOnlyFollowingDatabase
+```
+
+This is correct according to our rules for OneOf, but it doesn't work for ARM Spec generation due to the lack of a Name.
+
+#### Other factors
+
+We also have an `AzureName` property for many our custom resources, to allow the name in Azure to differ from the name in the cluster, due to different naming rules in each environment.
+
+## Option 1: Do Nothing
+
+Accept that the current decisions mean that we have to decline to support the `Microsoft.Kusto` resources, and any future resources that use this pattern.
+
+### Pros
+
+* Easy
+
+### Cons
+
+* Distasteful to decline to support a resource due to a technical limitation.
+* No guarantee that Kusto will be the only affected resource.
+* Risk that resources we already support will use this pattern in a new API version, rendering us unable to upgrade.
+
+## Option 2: Loosen the rules on Name
```

Comment:
I think I tend towards this option, which is awkward/weird but may make the most sense...

Resulting usage would be something like:
```
Database:
  spec:
    readWrite:
      name: ...
      azureName: ...
```

I think most usage of `Name` and `AzureName` by the generator is done through interfaces so as long as the `Database` type still implemented `AzureName()` we'd be OK? Making that function generate correctly will take some generator work though.

It _is_ a bit confusing for users but I'd argue it's about the same amount of confusing as using the Go SDK type (which takes an interface rather than a concrete type and thus requires the user to figure out what types impl that interface and then pick one)